lib/asn1.ml

```diff
@@ -28,17 +28,36 @@ struct
     }
     [@@deriving ord,eq,show]
 
+    let fix_negative_rsa = function
+      | Ok {n; e} ->
+        if Z.gt n Z.zero then
+          Ok { n; e }
+        else
+          (* A few tools generate keys without leading 0 bit, which gets interpreted
+             as a negative integer. This fix should account for these bad keys. *)
+          let length = n |> Z.numbits |> (fun n -> (n + 3) / 4) in
+          let all_fs = String.make length 'f' |> Z.of_string_base 16 in
+          let positive_n = Z.add all_fs n in
```

Comment:
How about putting a warning in the logs whenever we enter this branch?

Reply:
I see two ways to do that but I'm afraid none of them are very satisfactory: Either output to standard output/error, or output to a logging system.  The first is undesirable for a library and the second requires a dependency on a logging library.

That's why I think I prefer dropping the calculation and returning `Error` which can be handled by users whichever way they want.

Reply:
Yes, that's maybe the right thing to do. We know that this is rare.

I don't remember what happened before. Did the key parser crash on the key with the negative `n`? Ah no, OK I see. It just happily parsed it, produced a key with a negative `n` and something of ours choked on it further down the pipeline. So if key-parsers returned an `Error "key with negative n"` or something like that, we'd be in good shape.